cleanup: print warning message only if the function does not finish within 30 seconds

jvanz opened the pull request:
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

If a pod has a configmap/secret volume an annoying message shows up in the log approximately every 70 seconds. This happens because the desiredStateOfWorldPopulator sync loop always call the MarkRemountRequired. The function finds the volume plugin and check if the plugin requires mount. Configmap and secret plugins always returns true for that. Thus, the
reconciler code of the volume manager remounts the volume every time. This commit decrease the log level of that message  in the mount function from warning to V4.

**Which issue(s) this PR fixes**:

Fixes #90293 

**Does this PR introduce a user-facing change?**:

```release-note
Fixes the flooding warning messages about setting volume ownership for configmap/secret volumes
```

jvanz committed:
cleanup: decrease log level from warn to v3

If a pod has a configmap/secret volume an annoying message shows up
in the log approximately every 70 seconds. This happens because the
desiredStateOfWorldPopulator sync loop always call the
MarkRemountRequired. The function finds the volume plugin and check
if the plugin requires mount. Configmap and secret plugins always
returns true for that. Thus, the reconciler code of the volume manager
remounts the volume every time. This commit decrease the log level of
that message in the mount function from warning to V4.

Signed-off-by: José Guilherme Vanz <jguilhermevanz@suse.com>

jvanz commented:
/sig node

/sig storage

SergeyKanzhelev commented:
For the reference, this is PR where the warning was introduced: https://github.com/kubernetes/kubernetes/pull/83426/files

SergeyKanzhelev commented on pkg/volume/volume_linux.go:
```diff
@@ -46,7 +46,7 @@ func SetVolumeOwnership(mounter Mounter, fsGroup *int64, fsGroupChangePolicy *v1
-	klog.Warningf("Setting volume ownership for %s and fsGroup set. If the volume has a lot of files then setting volume ownership could be slow, see https://github.com/kubernetes/kubernetes/issues/69699", mounter.GetPath())
+	klog.V(3).Infof("Setting volume ownership for %s and fsGroup set. If the volume has a lot of files then setting volume ownership could be slow, see https://github.com/kubernetes/kubernetes/issues/69699", mounter.GetPath())
 
 	// This code exists for legacy purposes, so as old behaviour is entirely preserved when feature gate is disabled
```
nit: since you are making change in this file, can you please also fix this spelling issue:

```suggestion
	// This code exists for legacy purposes, so as old behavior is entirely preserved when feature gate is disabled
```

php-coder replied:
I'm not sure that it worth to be fixed BTW: https://ell.stackexchange.com/questions/202548/what-is-the-difference-between-behavior-and-behaviour

SergeyKanzhelev replied:
thanks for the info. Recommended based on VSCode spell checker =) I guess it uses American English than

jvanz replied:
Ok, marking this as resolved.

SergeyKanzhelev commented on pkg/volume/volume_linux.go:
```diff
@@ -46,7 +46,7 @@ func SetVolumeOwnership(mounter Mounter, fsGroup *int64, fsGroupChangePolicy *v1
 
 	fsGroupPolicyEnabled := utilfeature.DefaultFeatureGate.Enabled(features.ConfigurableFSGroupPolicy)
 
-	klog.Warningf("Setting volume ownership for %s and fsGroup set. If the volume has a lot of files then setting volume ownership could be slow, see https://github.com/kubernetes/kubernetes/issues/69699", mounter.GetPath())
```
I wonder if it would be better to log this after certain timeout passed instead of decreasing log level

jvanz replied:
Thanks! This is my very first PR and I have a question about what's this actually means. Do you believe it's better to have a timeout mechanism witch will print the warning message after `x`  time if this function is called (even if it is called multiples times) or something like that? If so, is there already a mechanism in place to do something like that?

SergeyKanzhelev replied:
I'm new too. I looked over the PR introduced this and the related issue and it looks like something that makes many people wonder why the pod has failed to start while it was mounting. Changing the log level will only help troubleshoot the issue when you can repro it consistently and you will need to change the logging level for kubelet. Thus I was thinking if this is really the issue that hard to troubleshoot, adding timeout might be a better solution. I haven't seen this pattern in k8s codebase yet.

jsafrane replied:
Logging the message after X seconds if the fsgroup change has not finished yet is actually a good idea. @gnufied, what do you think?

jsafrane replied:
Alternatively, SetVolumeOwnership could get a parameter that says how noisy it can be - it *should* log the message at higher level for "real" persistent storage (it can be slow) and not log anything for secrets/configmaps.

SergeyKanzhelev replied:
@kubernetes/sig-instrumentation-misc do you have a recommendation on this? Is there pattern to follow on reporting warning in case the method execution takes too long?

serathius replied:
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/logging.md

jvanz replied:
I've updated the code to print the warning only after a timeout. I do not know if the 30 seconds is the adequate time period. Let me know if I should change it.

jvanz committed:
Revert "cleanup: decrease log level from warn to v3"

This reverts commit ebece49936e635f151fdd8a64fa2b77fd183e817.

jvanz committed:
cleanup: print warning message after timeout

If a pod has a configmap/secret volume an annoying message shows up
in the log approximately every 70 seconds. This happens because the
desiredStateOfWorldPopulator sync loop always call the
MarkRemountRequired. The function finds the volume plugin and check
if the plugin requires mount. Configmap and secret plugins always
returns true for that. Thus, the reconciler code of the volume manager
remounts the volume every time. This commit change the SetVolumeOwnership
to print the warning only if the function does not finish within 30
seconds.

Signed-off-by: José Guilherme Vanz <jguilhermevanz@suse.com>

Klaven commented on pkg/volume/volume_linux.go:
```diff
@@ -46,23 +47,29 @@ func SetVolumeOwnership(mounter Mounter, fsGroup *int64, fsGroupChangePolicy *v1
 
 	return walkDeep(mounter.GetPath(), func(path string, info os.FileInfo, err error) error {
 		if err != nil {
+			timer.Stop()
```
I'm not sure you want to call timer.Stop() every time the func gets called. I think the intention here was to call it when the `walkDeep` object was done doing whatever it is that it does. It might be that you wanted to give `walkDeep` a reference too the timer so it can call stop when it's done.

jvanz replied:
Yeah, this is wrong. Thanks for bring my attention for that. I just update the PR removing all the calls and adding a just single `defer` call.

Klaven reviewed (commented):
First, welcome too kubernetes! And thanks for putting in the work to get this far! Here are some of my thoughts.

from my point of view this is not a terribly clean solution with a maintenance burden (`timer.Stop()` must be called on return) And this can't be solved by a `defer` because it has to be called outside the scope of this function.

In my opinion it might make since to log this in the function that calls `SetVolumeOwnership()` This puts the responsibility in the hands of the function that makes this call to know when it is done as `SetVolumeOwnership` does not know when it is done because it returns a function.

That being said a flag could also be used instead of a timer.

I personally would also like to see a test function that ensures the timer is eventually stopped, because this is the type of bug you could have a regression with and not even know/see it.

Klaven commented on pkg/volume/volume_linux.go:
```diff
@@ -46,7 +47,10 @@ func SetVolumeOwnership(mounter Mounter, fsGroup *int64, fsGroupChangePolicy *v1
+	timer := time.AfterFunc(30*time.Second, func() {
+		klog.Warningf("Setting volume ownership for %s and fsGroup set. If the volume has a lot of files then setting volume ownership could be slow, see https://github.com/kubernetes/kubernetes/issues/69699", mounter.GetPath())
+	})
+	defer timer.Stop()
```
I stand corrected. defer should work as `walkDeep` is executed before returning. meaning defer would only happen after. this lgtm now.

jvanz replied:
Thanks! :)

Klaven reviewed (commented):
defer looks good here I was wrong as I thought walkDeep was returned but it is executed. this is a clean solution. looks good to me.

SergeyKanzhelev commented:
/ok-to-test
/lgtm

liggitt commented:
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/92878/pull-kubernetes-node-e2e/1285390922829271040 flake is #93210

jsafrane commented:
/retest
/assign @gnufied 
/lgtm

jvanz commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

gnufied commented:
/lgtm

gnufied commented:
/approve

php-coder commented:
/retest

jvanz commented:
Should we set the 1.19 milestone?

ehashman reviewed (approved):
/approve

I've also seen this in 1.17, agreed it's annoying.

ehashman commented:
/test pull-kubernetes-e2e-kind-ipv6

ehashman commented:
/milestone v1.20

We're deep into the 1.19 code freeze, this can wait until thaw

ehashman commented:
/milestone v1.20

liggitt commented:
this will be moved back into the milestone as part of the [phased reopen for 1.20 merges](https://docs.google.com/document/d/1yI7o7R2bOB7QtkHUFMHMb5XObiJmzIAnJPuxhdAX-CA/edit)

msschl commented:
Can we get this as patch for v1.18.9?

jvanz commented:
> Can we get this as patch for v1.18.9?

I believe it will be easy to back port the change. It's quite simple. I just do not know what is the process for that and if it will be accepted.

ReillyTevera commented:
@jvanz [This](https://kubernetes.io/docs/contribute/generate-ref-docs/contribute-upstream/#cherry-picking-your-commit-into-a-release-branch) is the document you need to read. 

Note that the standards for getting into release branches that are already deep into the point releases is higher and is pretty much limited to functionality bugfixes. This is a more cosmetic change and might not be accepted (but it depends on the 1.18 team reviewer you get so I'd still recommend giving it a try). You'd probably have an easier time getting it into 1.19.

ehashman commented:
Backport applies cleanly to 1.18/1.19 but not to 1.17 (where this was apparently introduced). Opened PRs.

ehashman commented:
If someone wants to give the 1.17 backport a swing, feel free, but idk if we will get approvals :)

msschl commented:
A huge thanks to all participants especially to @jvanz @ehashman for your work. Thanks a lot 👍
